### ISSUE TITLE
Separate run defined commands from ad-hoc execution

### ISSUE THREAD

**Issue by AlbertDeFusco** [conda-incubator/conda-project#67]
Currently #38 utilizes `conda project run [command] [args ...]` to run either a defined command or any executable in the path of the activated env in your shell.

The `run` command should instead only execute defined commands, with optional arguments. A separate command, perhaps `exec` will support ad-hoc, one-time execution for undefined commands

**Comment by maximlt** [comment]
I had a look at similar projects and the API they offer. Not that this was particularly fun to do :) but I tend to think that there's so much prior art on this that it makes sense to look around and maybe converge on a solution that has already been adopted elsewhere to ease the adoption of conda-project. I chose to look at poetry, hatch, pdm and tox. You'll note these are Python projects only.

### TL;DR

- poetry, hatch and pdm have a `run` command that allows to run arbitrary commands.
- poetry with poethepoet plugin, hatch and pdm can execute named commands in environments via their `run` command. hatch and pdm give precedence to names commands over arbitrary commands when they clash (this is irrelevant for poetry as the poethepoet plugin does not allow to run arbitrary commands).
- tox's `run` command is limited to execute non-arbitrary commands in environments, the new `tox exec` command can be used to run arbitrary commands in specific environments.
- Only poethepoet allows to pass named arguments to the CLI
- Substituting arguments is done by templating the command with `{args:default}` and substituting that with whatever is passed after the command (hatch, pdm) or whatever is passed after `--` (tox).

### poetry

[poetry run](https://python-poetry.org/docs/cli/#run) runs the given command inside the project’s virtualenv. The example in their doc is `poetry run python -V`.

poetry can also run scripts (entry points) defined as such in a `pyproject.toml` file and executed with `poetry run my_entrypoint`.

```toml
[tool.poetry.scripts]
my_entrypoint = 'my_package.module:main'
```

poetry doesn't support more than that directly when it comes to executing tasks, there are however plugins that take care of it. I chose to focus on [poethepoet](https://github.com/nat-n/poethepoet).

This is basically a pretty advanced task runner, whose configuration can be defined in a `pyproject.toml` file. It offers different kind of tasks (commands, scripts, shell, sequence, expression, switch, ref).

This is how *command* tasks are defined, that type of command being what we are focusing on in this issue (example taken from their README):

```toml
[tool.poe.tasks]
format = "black ."  # strings are interpreted as commands by default
clean = """
# Multiline commands including comments work too. Unescaped whitespace is ignored.
rm -rf .coverage
       .mypy_cache
       .pytest_cache
       dist
       ./**/__pycache__
"""
lint = { "cmd": "pylint poethepoet" }  # Inline tables with a cmd key work too
greet = "echo Hello $USER"  # Environment variables work, even though there's no shell!
```

CLI arguments can be defined per command, setting their name, default value, type, options, etc. In the example below I create the `my_echo` command with two CLI arguments, you'd invoke it with `poetry poe my_echo --host domain --port 5006`.

```toml
[tool.poe.tasks.my_echo]
cmd = "echo ${host} ${port}"
args = [{ name = "host", default = "localhost" }, { name = "port", default = "9000" }]
```

An advantage of defining the inputs this way is that they can be inspected via the command line. This is a part of the output of `poetry poe`:

```
CONFIGURED TASKS
  my_echo
    --host
    --port
```

Also of interest for this discussion, extra arguments passed to the CLI following the task name are appended to the end of a cmd task. So calling the `my_python` command below with `poetry poe my_python --version` will call `python --version`.

```toml
[tool.poe.tasks]
my_python = "python"
```

From what I have seen, it is not possible to combine passing extra commands together with named arguments to the CLI.

### hatch

[hatch run](https://hatch.pypa.io/latest/environment/) works the same as poetry's `run`, it executes the command in an environment as if was activated. Their default example is `hatch run python -c "import sys;print(sys.executable)"`.

`hatch run` is not limited to that though, it can also execute *scripts*, which are really command tasks. The following bit shows how to add a `bar` command to the *default* environment, that is then executed running `hatch run bar`.

```toml
[tool.hatch.envs.default.scripts]
bar = "echo bar"
```

hatch can handle multiple environments and have scripts linked to environments, the `cov` command/script below can be executed running either `hatch run tests:cov` or `hatch env run --env tests cov`.

```toml
[tool.hatch.envs.tests]
dependencies = [
  "pytest",
]
[tool.hatch.envs.tests.scripts]
cov = "pytest --cov-report=term-missing --cov-config=pyproject.toml --cov=hatch_demo --cov=tests {args}"
```

I have not found a way to pass named CLI arguments to hatch, the only option to extend a script being to use the `{args}` template with an optional default value `{args:default`} as in `hello = "echo {args:hello}"`.

Named scripts take precedence over arbitrary commands. For example, `hatch run python` will run the `python` script (`python = "echo bar"`) instead of running the Python executable.

### pdm

[pdm run](https://pdm.fming.dev/latest/usage/scripts/#arbitrary-scripts) can run arbitrary scripts. The example in their docs is `pdm run flask run -p 54321`.

It can also run *scripts* defined in the `pyproject.toml` file, executed with `pdm run <scriptname>`.

```toml
[tool.pdm.scripts]
hello = "echo hello"
```

Any following arguments will be appended to the command, so executing `pdm run hello bob` would print `hello bob` in the console.

I haven't seen any way to pass named CLI arguments.

### tox

`tox` doesn't really advertise itself as a project management tool like the previous tools, it focuses instead on testing, but its features are anyway pretty close so it's interesting to look at how its API.

`tox run` is used to run the commands of environments. Additional positional arguments can be passed to a command if it defines `{posargs}` or `{posargs:default}`, by passing them after `--`. In the example below running `tox run -e py310` will display `hello` in the console, and running `tox run -e py310 -- hey` will display `hey`.

```ini
[tox]
env_list = py310

[testenv]
allowlist_externals = echo
commands =
    python -c "print('{posargs: hello}')"
```

`tox exec` is used to run arbitrary commands in an environment, for example running `tox exec -e py310 python --version`. The `exec` commands was added in the recently released [tox 4](https://tox.wiki/en/latest/faq.html#new-features-in-tox-4), quoting their docs: `We redesigned our CLI interface, we no longer try to squeeze everything under single command, instead now we have multiple sub-commands`.

**Comment by maximlt** [comment]
Regarding whether `conda project run` should be allowed to execute arbitrary commands, the summary above shows that's a feature offered by poetry, hatch and pdm and as such would make sense for conda-project. The new `tox exec` command goes in favor of having a `conda project exec` command dedicated to that.

One pretty important point that was not mentioned above is that the **existing `conda run` command is used to execute arbitrary commands**. It seems to me that this is enough to allow `conda project run` to execute arbitrary commands, isn't it?

**Comment by AlbertDeFusco** [comment]
So would you then recommend `run` for ad-hoc commands and `exec` for defined commands?

**Comment by maximlt** [comment]
As for templating and expanding commands via the CLI, I haven't seen an approach that covers both of them. The approach I liked wrt templating is the one from poethepoet that allows to define named CLI arguments. I've been using `doit` a lot lately and it has the same feature, I found it very pleasant to work with a task runner whose tasks and parameters can be inspected via the CLI (e.g. `doit help command_name`). Wrt expanding, having to separate these extra arguments with `--` as tox does made sense to me.

I'm not sure to which extent this is possible but merging these two concepts this would look like small example. A command could be defined with names arguments (here `--port`).

```yaml
dashboard:
  cmd: panel serve app.py --port ${port}
  args:
  - name: port
  - default: 5006
  - type: int
  - description: The port to serve the app to.
```

And it would be possible to call it passing additional arguments provided after `--`:

    conda project run dashboard --port 5007 -- --show

But to be honest, this needs a little more thinking :) Running tasks is certainly a core feature of `conda-project` and similar tools provide pretty advanced features like different type of tasks (commands, shells, composite tasks) and more. I can already sense the influx of feature requests coming in!

**Comment by maximlt** [comment]
> So would you then recommend `run` for ad-hoc commands and `exec` for defined commands?

Seems to me having a single `run` command is fine, being both in line with the other Python project management tools and `conda` itself. I'm not particularly worried about nameclashes.

**Comment by jbednar** [comment]
The above discussion all sounds very reasonable to me. I'm inclined to support `run` for both arbitrary commands (to match conda) and for predefined commands, particularly if both arbitrary and predefined ones will accept additional arguments. I wouldn't object to keeping them separate, but haven't ever noticed any particular issue with having them combined in `anaconda-project`.

**Comment by AlbertDeFusco** [comment]
Thank you for the input.

**Comment by AlbertDeFusco** [comment]
thanks again, all. For #73 `run` will support both defined and ad-hoc commands with no `--` spacer for arguments. Defined arguments take precedence on collision. Let's separately talk about templating and defined parameters.
